source/common/upstream/cluster_manager_impl.cc

```diff
@@ -931,14 +932,22 @@ void ClusterManagerImpl::postThreadLocalClusterUpdate(const Cluster& cluster, ui
                                                       const HostVector& hosts_removed) {
   const auto& host_set = cluster.prioritySet().hostSetsPerPriority()[priority];
 
-  tls_.runOnAllThreads([name = cluster.info()->name(), priority,
-                        update_params = HostSetImpl::updateHostsParams(*host_set),
-                        locality_weights = host_set->localityWeights(), hosts_added, hosts_removed,
-                        overprovisioning_factor = host_set->overprovisioningFactor()](
-                           OptRef<ThreadLocalClusterManagerImpl> cluster_manager) {
-    cluster_manager->updateClusterMembership(name, priority, update_params, locality_weights,
-                                             hosts_added, hosts_removed, overprovisioning_factor);
-  });
+  tls_.runOnAllThreads(
+      [name = cluster.info()->name(), priority,
+       update_params = HostSetImpl::updateHostsParams(*host_set),
+       locality_weights = host_set->localityWeights(), hosts_added, hosts_removed,
+       overprovisioning_factor = host_set->overprovisioningFactor()](
+          OptRef<ThreadLocalClusterManagerImpl> cluster_manager) {
+        cluster_manager->updateClusterMembership(name, priority, update_params, locality_weights,
+                                                 hosts_added, hosts_removed,
+                                                 overprovisioning_factor);
+      },
+      // This completion is guaranteed to be executed on master thread.
+      // Still the host can be snapped by dataplane and need to addressed separately.
+      [hosts_added_guard = hosts_added, hosts_removed_guard = hosts_removed]() {
```

Comment:
Why is this needed? If it is needed this is going to cause extra host copies on every update and we probably should be figuring out how to turn this into a shared_ptr. I noticed this on my other PR that I recently did in this area (you need a main merge to pick that up)

Reply:
Host copy is much more expensive and unnecessary if we choose to post clusterInfo(equivalently,  owner of clusterinfo). 

Deleted.